[3D feature] Skybox feature

NEDJIMAbelgacem opened the pull request:
## Skybox rendering
QGIS 3D views support rendering skyboxes.
From the the 3D configuration dialog you can enable the skybox rendering and choose from 2 different ways of providing skyboxes.
### Different textures for each face:
![distinct_faces](https://user-images.githubusercontent.com/29183781/89114381-9be49d80-d473-11ea-9d00-197dc4a6351a.png)
Specify each texture seperately according to which face it belongs to.
### High resolution 360 image:
![360Image](https://user-images.githubusercontent.com/29183781/89114388-ad2daa00-d473-11ea-9281-e238babd6596.png)
You can use 360 images like .hdr ones from [hdrihaven](https://hdrihaven.com/)
### GIFs of the skybox rendering:
![skybox1](https://user-images.githubusercontent.com/29183781/89114459-802dc700-d474-11ea-842c-f27985f6f736.gif)
![skybox2](https://user-images.githubusercontent.com/29183781/89114463-8459e480-d474-11ea-82ba-48a270bed5d3.gif)

NEDJIMAbelgacem committed:
Added skybox rendering settings widget

NEDJIMAbelgacem committed:
skybox rendering initial implementation

NEDJIMAbelgacem committed:
skybox settings class and related slots/signals

NEDJIMAbelgacem committed:
Get cube map textures loading into seperate classes

NEDJIMAbelgacem committed:
HDR skyboxes implementation

NEDJIMAbelgacem committed:
Split skybox class into 2 classes for HDR and 6 faces

NEDJIMAbelgacem committed:
skybox settings UI

NEDJIMAbelgacem committed:
Persist skybox settings across settings dialog creations

NEDJIMAbelgacem committed:
Delete previous skybox implementation and save skybox settings as a DOM element

NEDJIMAbelgacem committed:
Load skybox using skybox settings when project is opened

NEDJIMAbelgacem committed:
Fix skybox cubemap issue

NEDJIMAbelgacem committed:
def_window_title_check test fix

NEDJIMAbelgacem committed:
Add licence

NEDJIMAbelgacem committed:
initialize mSkyboxSettings in Qgs3DMapSettings constructor

NEDJIMAbelgacem committed:
Add docs

NEDJIMAbelgacem committed:
Remove 3D canvas from Qgs3DMapConfigWidget constructor

NEDJIMAbelgacem committed:
Fix docs

NEDJIMAbelgacem committed:
Merge branch 'master' into skybox-feature

NEDJIMAbelgacem committed:
Use the same vertex shader for different skybox implementations

NEDJIMAbelgacem committed:
Fix merge conflict

NEDJIMAbelgacem committed:
Add _3D_EXPORT

nyalldawson commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -325,7 +320,6 @@ void Qgs3DMapScene::onCameraChanged()
 
 void Qgs3DMapScene::updateScene()
 {
-  QgsEventTracing::addEvent( QgsEventTracing::Instant, QStringLiteral( "3D" ), QStringLiteral( "Update Scene" ) );
```
What's the rationale behind removing this?

NEDJIMAbelgacem replied:
Ah that was just a random mistake. Sorry about that

nyalldawson commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -831,6 +825,10 @@ void Qgs3DMapScene::addCameraViewCenterEntity( Qt3DRender::QCamera *camera )
   {
     mEntityCameraViewCenter->setEnabled( mMap.showCameraViewCenter() );
   } );
+//  connect( &mMap, &Qgs3DMapSettings, this, [this]
```
remove

nyalldawson commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -861,6 +859,40 @@ void Qgs3DMapScene::updateSceneState()
 
+void Qgs3DMapScene::onSkyboxSettingsChanged()
+{
+  qDebug() << __FUNCTION__;
```
remove

nyalldawson commented on src/3d/qgsskyboxentity.cpp:
```diff
@@ -0,0 +1,193 @@
+  mMesh->setXZMeshResolution( QSize( 2, 2 ) );
+  mMesh->setYZMeshResolution( QSize( 2, 2 ) );
+
+//   TODO: change the kybox position according to camera
```
```suggestion
//   TODO: change the skybox position according to camera
```

NEDJIMAbelgacem committed:
Move writeXml and readXml to qgsskyboxsettings.cpp

nyalldawson commented on src/3d/qgsskyboxentity.cpp:
```diff
@@ -0,0 +1,193 @@
+
+void QgsHDRSkyboxEntity::reloadTexture()
+{
+  mLoadedTexture->setSource( QUrl::fromUserInput( mHDRTexturePath ) );
```
This will need to use QgsImageCache, just like the textures in the phong material (otherwise embedded textures won't work)

NEDJIMAbelgacem replied:
I tried to do this for HDR images but apparently that's not supported with QImage.

nyalldawson commented on src/3d/qgsskyboxentity.h:
```diff
@@ -0,0 +1,106 @@
+    QgsSkyboxEntity( QNode *parent = nullptr );
+
+  protected:
+    Qt3DRender::QEffect *mEffect;
```
initialize all these to nullptr here

nyalldawson commented on src/3d/qgsskyboxentity.h:
```diff
@@ -0,0 +1,106 @@
+    QgsHDRSkyboxEntity( const QString &hdrTexturePath, Qt3DCore::QNode *parent = nullptr );
+
+    //! Returns the path of the current texture in use
+    QString getHDRTexturePath() const { return mHDRTexturePath; }
```
```suggestion
    QString hdrTexturePath() const { return mHDRTexturePath; }
```

nyalldawson commented on src/3d/qgsskyboxsettings.h:
```diff
@@ -0,0 +1,127 @@
+
+    /**
+     * Sets the type of the skybox
+     * the type can be one of 3 types: "Textures collection", "HDR texture", "Distinct Faces"
```
It would be cleaner to use an enum for these types, instead of the string values

nyalldawson commented on src/3d/qgsskyboxsettings.h:
```diff
@@ -0,0 +1,127 @@
+    //! Returns the HDR texture path of a skybox of type "HDR texture"
+    QString hdrTexturePath() const { return mHDRTexturePath; }
+    //! Sets the HDR texture path of a skybox of type "HDR texture"
+    void setHDRTexturePath( const QString &texturePath ) { mHDRTexturePath = texturePath; }
```
```suggestion
    void setHdrTexturePath( const QString &texturePath ) { mHDRTexturePath = texturePath; }
```

nyalldawson reviewed (commented):
Nice!

NEDJIMAbelgacem committed:
Refactoring

NEDJIMAbelgacem committed:
Add enum skybox types

NEDJIMAbelgacem commented:
Thanks for the reviews @nyalldawson !

NEDJIMAbelgacem committed:
Add SIP_NO_FILE and fix bindings

NEDJIMAbelgacem committed:
Fix merge conflict
Merge branch 'master' into skybox-feature

NEDJIMAbelgacem committed:
Add missing documentation

NEDJIMAbelgacem commented:
Hi @DelazJ , what documentations I need to add ?

DelazJ commented:
Hi @NEDJIMAbelgacem 

QGIS has a user-oriented documentation at https://docs.qgis.org, and some 3D stuffs at https://docs.qgis.org/testing/en/docs/user_manual/introduction/qgis_gui.html#d-map-view. The "Needs Documentation" label is meant to generate, once this PR is merged, an issue report in the [docs repository](www.github.com/qgis/QGIS-Documentation/issues) with full content of your first message, so that writers are aware of this new feature and document it. So the clearer and up to date the first message of the PR is (and i feel it is the case here), the most helpful it's for docs team (which btw welcomes anyone willing to help write about _their_ features... :wink: ).
Thanks for the contributions.

NEDJIMAbelgacem commented:
Ah I see, I wrote the PR message from "what I changed" point of view. I will write more user oriented message for this PR and my other PRs as well. Thanks @DelazJ

NEDJIMAbelgacem committed:
Fix compilation error for ubuntu system packages users

wonder-sk reviewed (commented):
Great job - the code is looking very good!

I have a few suggestions:
- we can probably safely get rid of TexturesCollectionSkybox type... it is essentially the same thing as DistinctTexturesSkybox just with fixed naming of files, and I have not really seen that particular naming scheme (chosen in the original qt3dextras implementation) used anywhere else - that's why I would prefer to get rid of it and make both GUI and code cleaner
- can we rename HDRSkybox to "Panoramic" skybox (both in the code and in GUI)? I think in this case it does not really matter whether the images are HDR (high dynamic range) or not - what really matters that they should be panoramic. This would be also closer to how this format is called elsewhere (https://docs.unity3d.com/Manual/skybox-shaders.html). I guess it is just a coincidence that the skybox panoramas you were using were also HDR images :-)
- it may be interesting to have support for "6 sides" type skybox (as called in Unity) as it seems that some people/engines use this format - e.g. http://scmapdb.com/wad:skybox-skies
- naming of files in GUI for Distinct Textures Skybox (Neg Y, Pos X, ...) could be a bit more user friendly - e.g. Right (+X), Left (-X), Top (+Y), ...

NEDJIMAbelgacem commented on src/3d/qgs3dmapsettings.h:
```diff
@@ -426,6 +412,16 @@ class _3D_EXPORT Qgs3DMapSettings : public QObject, public QgsTemporalRangeObjec
+    /**
+     * Returns the current configuration of the skybox
+     */
+    QgsSkyboxSettings skyboxSettings() const SIP_SKIP { return mSkyboxSettings; }
```
It's because when I changed the skybox type to being an enum I both didn't know how to correctly export it and the enum was inside QgsSkyboxEntity class that inherits from Qt3DCore::QEntity which wasn't available for python bindings.

NEDJIMAbelgacem committed:
Use path resolver's read/writePath()
Remove QSkyboxEntity include
Use deleteLater() instead of delete
Remove unneeded todo
Add \since to skyboxSettings in qgs3dmapsettings.h

NEDJIMAbelgacem committed:
Rename HDR skybox to panoramic skybox

NEDJIMAbelgacem committed:
get rid of TexturesCollectionSkybox

NEDJIMAbelgacem committed:
Add tr to skybox types

NEDJIMAbelgacem committed:
Use currentIndex instead of currentText for combobox

NEDJIMAbelgacem committed:
Remove transform and unneeded includes from qskyboxentity.h

NEDJIMAbelgacem commented on src/3d/qgsskyboxentity.cpp:
```diff
@@ -0,0 +1,199 @@
+QgsHDRSkyboxEntity::QgsHDRSkyboxEntity( const QString &hdrTexturePath, QNode *parent )
+  : QgsSkyboxEntity( parent )
+  , mHDRTexturePath( hdrTexturePath )
+  , mLoadedTexture( new Qt3DRender::QTextureLoader( parent ) )
```
I followed what Qt3D does. I tried using Qt3DRender::QTextureImage but it didn't work properly and opengl was using a random texture from the scene.

NEDJIMAbelgacem commented on src/3d/qgsskyboxentity.cpp:
```diff
@@ -0,0 +1,199 @@
+  mMesh->setXZMeshResolution( QSize( 2, 2 ) );
+  mMesh->setYZMeshResolution( QSize( 2, 2 ) );
+
+  //   TODO: change the skybox position according to camera
```
No it's not. The skybox cube position shouldn't matter because it gets removed with the vertex shader.

NEDJIMAbelgacem committed:
Change skybox faces to more intuitive names in the UI

NEDJIMAbelgacem committed:
use qgis::overload instead of QOverload

NEDJIMAbelgacem commented:
Thanks for the reviews @wonder-sk !

DelazJ commented:
> Ah I see, I wrote the PR message from "what I changed" point of view. I will write more user oriented message for this PR and my other PRs as well.

@NEDJIMAbelgacem It would be nice, indeed. Thanks. And it would also help since this description will be picked to generate the release changelog.
Otherwise, I opened some 3D docs PR at https://github.com/qgis/QGIS-Documentation/pull/5994 and https://github.com/qgis/QGIS-Documentation/pull/5993 and I'm not quite confident, so if ever you want to give it a look... Thanks

Saijin-Naib commented:
> Great job - the code is looking very good!
> I have a few suggestions:
> it may be interesting to have support for "6 sides" type skybox (as called in Unity) as it seems that some people/engines use this format - e.g. http://scmapdb.com/wad:skybox-skies

I raised this as a Feature Request in Issue #32981

waytgibbs commented:
In attempting to use the new Skybox feature with a panoramic HDR background in the 9/15/2020 nightly build, I'm observing rendering issues when exporting animation frames from a keyframed flyover sequence in the 3D map view. See attached images for a four-frame sequence that illustrates the issue, which occurs at irregular intervals. These glitches are not visible when stepping through the flyover interactively within QGIS.
![HiRes51383](https://user-images.githubusercontent.com/6627581/93371494-530c5c80-f807-11ea-87dc-ade8c7958e8c.jpg)
![HiRes51384](https://user-images.githubusercontent.com/6627581/93371512-57387a00-f807-11ea-8a6b-0bfa40a088a1.jpg)
![HiRes51385](https://user-images.githubusercontent.com/6627581/93371515-59023d80-f807-11ea-8928-ce46e85b96d2.jpg)
![HiRes51386](https://user-images.githubusercontent.com/6627581/93371522-5a336a80-f807-11ea-84ed-30abe0dba5e2.jpg)
